registry/handlers/app.go

```diff
@@ -184,6 +184,18 @@ func NewApp(ctx context.Context, config *configuration.Configuration) *App {
 		}
 	}
 
+	// configure tag lookup concurrency limit
+	if p := config.Storage.TagParameters(); p != nil {
+		l, ok := p["concurrency_limit"]
+		if ok {
+			limit, ok := l.(int)
+			if !ok {
+				panic("tag lookup concurrency limit config key must have a integer value")
```

Comment:
What are the semantics of an explicit concurrency limit of `0`? `-1`?

Reply:
During the initialization of the tagStore, a judgment is made, and if it is greater than 0, it will be set; otherwise, the concurrency limit is set to the default value.
```golang
limit := DefaultConcurrencyLimit
if repo.tagLookupConcurrencyLimit > 0 {
	limit = repo.tagLookupConcurrencyLimit
}
```

Reply:
We should document that in the docs FWIW

Reply:
> We should document that in the docs FWIW

done

Reply:
Defining all non-positive integers to mean `GOMAXPROCS` does not leave any values to signify the other possible special case: unbounded concurrency / no limit. To be clear, **I don't think an option for unbounded concurrency should be added as part of this PR.** I simply want to reserve the negative numbers for future use by having it be an error to set `concurrencylimit` to a negative value.

Reply:
I suggest that we _do not handle_ unbounded concurrency. At least not as part of this PR. What I am suggesting is the following:

| `concurrencylimit` value | Result |
|--------|--------|
| (unset) | Tag lookups are limited to `GOMAXPROCS` concurrency (the default) |
| `0` | Tag lookups are limited to `GOMAXPROCS` concurrency (the default) |
| Positive integer, e.g. `3` | Tag lookups are limited to the value, e.g. `3` | 
| Negative integer, e.g. `-1` | Configuration error |

Making negative numbers an error today guarantees that no valid registry configuration can have a `concurrencylimit` option set to a negative number. That gives us the freedom to redefine negative values to mean something valid in the future _as a non-breaking change_. One potential _future_ definition of negative `concurrencylimit` values is to configure unbounded concurrency. But we don't need that today.

Note that setting `concurrencylimit: 0` should be allowed. It is very useful to have an "explicit default" value for distribution configuration as it allows the user to be explicit about their intent to use `GOMAXPROCS` as the concurrency limit in their configuration, or to override e.g. `concurrencylimit: 4` in the YAML config by setting the environment variable `REGISTRY_TAG_CONCURRENCYLIMIT=0`.

Reply:
Thanks for your nice explanation, I have updated it, PTAL.